---
title: "Data plan for geneExp"
author: "Bingfang"
date: "12/20/2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. Objective

Identify a list of true expressed genes of a specific cell type, which will use to for functional analysis of biological process/pathway of the cell type.

## 2. Set environment


```{r, echo=TRUE,warning=FALSE,message=FALSE}
#install.packages("sads")  # not working on local MAC
library(sads)
#install.packages("BiocParallel")
library(BiocParallel)
library(Seurat)
library(org.Hs.eg.db)
#install.packages("Matrix")
library(Matrix)
# package for goenrichment
library(clusterProfiler)
library(AnnotationDbi)

# Get gene-specific distributions
# Publication:https://academic.oup.com/biostatistics/article/23/4/1150/6622011
# https://github.com/igrabski/scRNAseq-cell-type

load('params_EM_81020.rda') 
pi.all.g <- params[[1]]         # π-null,π-off, probabilty of null and off state.    
g.on.all.g <- params[[2]]       # σ1j⁠
g.off.all.g <- params[[3]]      # σ0j
a.all.g <- params[[4]]          # αj⁠
sigma.all.g <- params[[5]]      # μ0j,μ1j shift from (μ0 , μ1)
mu.g<-c(-12.73153,-8.301349)    # global means for gene rates on a log scale, (μ0 , μ1)

```

## 2. Data collection and preprocess.

### 2.1 Read .rds file

HLCA data from cellXgene

cellRef data from cellXgene

The authors of cellRef match "celltype_level3_fullname" with "putative CL label" and "CL term"

```{r, results='hide'} 

hlca = readRDS("/gpfs/gsfs10/users/xubr/cellXgene/scRNAseq-cell-type/data/b351804c-293e-4aeb-9c4c-043db67f4540_hlca_core.rds")
cellRef = readRDS("/gpfs/gsfs10/users/xubr/cellXgene/scRNAseq-cell-type/data/443f7fb8-2a27-47c3-98f6-6a603c7a294e_cellRef.rds")


summary(hlca@assays$RNA@meta.features$feature_type)
head(hlca@assays$RNA@meta.features$feature_name)
table(hlca@assays$RNA@meta.features$feature_is_filtered)

summary(cellRef@assays$RNA@meta.features$feature_type)
head(cellRef@assays$RNA@meta.features$feature_name)
cellRef@assays$RNA@meta.features$origSymbol
table(cellRef@assays$RNA@meta.features$feature_is_filtered)
table(cellRef@meta.data$cell_type_ontology_term_id)
table(cellRef@meta.data$putative_CL_label)
table(cellRef@meta.data$celltype_level3_fullname)


```

IncRNA ensembl ID may not be stable
e.g. ENSG00000238009 
Ensembl gene ENSG00000238009 is no longer in the database.It has been mapped to 1 current identifier  ENSG00000241860.


### 2.2 remove low quality cells

```{}
## plot

jpeg(file="HLCA_nFeature.jpeg")
VlnPlot(hlca, features = c("nFeaturess_RNA_detected"))   #,raster=FALSE)
dev.off()

jpeg(file="cellRef_nFeature.jpeg")
VlnPlot(cellRef, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
dev.off()

## filter

## need to convert to gene symbol to filter percent.mt < 5
## pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

hlca_filtered <- subset(hlca, subset = nFeaturess_RNA_detected > 200 & nFeaturess_RNA_detected < 2500)
cellRef_filtered <- subset(cellRef,subset = nFeature_RNA > 200 & nFeature_RNA < 2500)
```

HLCA and cellRef are combined datasets, Likely, they should not be filtered using nFeature.

```{r}
## prior and after filter
hlca@assays$RNA@counts@Dim
hlca_filtered@assays$RNA@counts@Dim

cellRef$RNA@counts@Dim
cellRef_filtered@assays$RNA@counts@Dim
```

### 2.3 Subset by cell type

```{r}
unique(hlca$ann_finest_level)
unique(cellRef$celltype_level3_fullname)

hlca_Alveolar_macrophages <- subset(hlca, subset = ann_finest_level == "Alveolar macrophages")
hlca_Alveolar_fibroblasts <- subset(hlca, subset = ann_finest_level == "Alveolar fibroblasts")
hlca_2alveolar <- subset(hlca, subset = ann_finest_level %in% c("Alveolar macrophages", "Alveolar fibroblasts"))

cellRef_Alveolar_macrophage <- subset(cellRef, subset = celltype_level3_fullname == "Alveolar macrophage")
cellRef_Alveolar_fibroblast2 <- subset(cellRef, subset = celltype_level3_fullname == "Alveolar fibroblast 2")
cellRef_2alveolar <- subset(cellRef, subset = celltype_level3_fullname %in% c("Alveolar macrophage", "Alveolar fibroblast 2"))


dim(hlca_2alveolar$RNA)
dim(cellRef_2alveolar$RNA)
head(hlca_2alveolar$RNA@counts)
tail(cellRef_2alveolar$RNA@counts)


```


## 2.4 geneXcell matrix


```{r}
# hlca  is seurat object from readRDS
hlca_matrix_Alveolar_macrophages <- as.matrix(hlca_Alveolar_macrophages@assays$RNA@counts)
hlca_matrix_Alveolar_fibroblasts <- as.matrix(hlca_Alveolar_fibroblasts@assays$RNA@counts) 
hlca_matrix_2alveolar <- as.matrix(hlca_2alveolar@assays$RNA@counts)
hlca_matrix_2alveolar_labels<-c(rep('Alveolar macrophages',dim(hlca_matrix_Alveolar_macrophages)[2]),rep('Alveolar fibroblasts',dim(hlca_matrix_Alveolar_fibroblasts)[2]))

cellRef_matrix_Alveolar_macrophage <- as.matrix(cellRef_Alveolar_macrophage@assays$RNA@counts)
cellRef_matrix_Alveolar_fibroblast2 <- as.matrix(cellRef_Alveolar_fibroblast2@assays$RNA@counts) 
cellRef_matrix_2alveolar <- as.matrix(cellRef_2alveolar@assays$RNA@counts)
cellRef_matrix_2alveolar_labels<-c(rep('Alveolar macrophage',dim(cellRef_matrix_Alveolar_macrophage)[2]),rep('Alveolar fibroblast2',dim(cellRef_matrix_Alveolar_fibroblast2)[2]))

```

## 4. Calculate rate and probability



```{r, message=FALSE}

# Train reference data
trainReference <- function(ref,pi.all=pi.all.g,mu=mu.g,a=a.all.g,sigma.all=sigma.all.g,
                           g.on.all=g.on.all.g,g.off.all=g.off.all.g,discrim_only=FALSE) {
  # Subset to genes present in both reference data and knowledge base
  common <- intersect(names(ref),rownames(pi.all))
  if (discrim_only) {
    common <- intersect(common,discrim.g)
  }
  N <- sum(ref)
  ref <- ref[common]
  pi.all2 <- pi.all[common,]
  a.all2 <- a[common]
  sigma.all2 <- sigma.all[common,]
  g.on.all2 <- g.on.all[common]
  g.off.all2 <- g.off.all[common]
  
  # Compute mixing probability for each component
  prob.exp <- pi.all2[,1]*dnbinom(ref,1,1/((N/(a.all2))+1))
  prob.ln1 <- sapply(common,function(j) 
    pi.all2[j,2]*sads::dpoilog(ref[j],mu[1]+g.off.all2[j]+log(N),sigma.all2[j,1]))
  prob.ln2 <- sapply(common,function(j) 
    (1-pi.all2[j,1]-pi.all2[j,2])*sads::dpoilog(ref[j],mu[2]+g.on.all2[j]+log(N),sigma.all2[j,2])) 
  return(data.frame(rate=ref/N,exp=prob.exp/(prob.exp+prob.ln1+prob.ln2),ln1=prob.ln1/(prob.exp+prob.ln1+prob.ln2),ln2=prob.ln2/(prob.exp+prob.ln1+prob.ln2)))
}


# Train all reference
trainAllReference <- function(data,labels,pi.all=pi.all.g,mu=mu.g,a=a.all.g,sigma.all=sigma.all.g,
                              g.on.all=g.on.all.g,g.off.all=g.off.all.g,discrim_only=FALSE) {
  cell.types <- unique(labels)
  dataSums <- lapply(cell.types,function(c) rowSums(data[,labels==c]))

  d.list <- bplapply(dataSums,trainReference,pi.all=pi.all,mu=mu,a=a,sigma.all=sigma.all,
                     g.on.all=g.on.all,g.off.all=g.off.all,discrim_only=discrim_only)
  names(d.list) <- cell.types
  return(d.list)
}


hlca_2alveolar.d <- trainAllReference(hlca_matrix_2alveolar,hlca_matrix_2alveolar_labels,discrim_only=FALSE)
head(hlca_2alveolar.d[[1]])

cellRef_2alveolar.d <- trainAllReference(cellRef_matrix_2alveolar,cellRef_matrix_2alveolar_labels,discrim_only=FALSE)
head(cellRef_2alveolar.d[[1]])

```

## 5. getBarcode

```{r}
getBarcode<-function(d.list) {
  barcode <- sapply(d.list,function(x) 1-x[,2]-x[,3])      # ### Barcode is   ln2=prob.ln2/(prob.exp+prob.ln1+prob.ln2) if greater than 0
  barcode[barcode<0] <- 0
  colnames(barcode) <- names(d.list)
  rownames(barcode) <- rownames(d.list[[1]])
  return(barcode)
}

barcodes_hlca <- getBarcode(hlca_2alveolar.d)
barcodes_hlca<- as.data.frame(barcodes_hlca)
head(barcodes_hlca)

barcodes_cellRef <- getBarcode(cellRef_2alveolar.d)
barcodes_cellRef<- as.data.frame(barcodes_cellRef)
head(barcodes_cellRef)
```

```{r}
for (cut_off in c(0.75,0.95)) {
  cat("The following is the length of the expressed gene list for cell type Alve, and ionocyte, if the probability of gene expression is great than ", cut_off , "\n")
  
  mucus_secreting_geneList <-rownames(barcodes)[barcodes$`mucus_secreting` > cut_off]
  cat("mucus_secreting:", length(mucus_secreting_geneList), "\n")
  write.table(mucus_secreting_geneList, file = paste0("mucus_secreting", cut_off, ".tsv"), sep = "\n", row.names = FALSE, col.names = FALSE, quote = FALSE)
  ionocyte_geneList <-rownames(barcodes)[barcodes$`ionocyte` > cut_off]
  cat("ionocyte:", length(ionocyte_geneList), "\n")
  write.table(ionocyte_geneList, file = paste0("ionocyte", cut_off, ".tsv"), sep = "\n", row.names = FALSE, col.names = FALSE, quote = FALSE)
  
}
```

## 6. Visualization

```{r}
library(clusterProfiler)
mucus_secreting_geneList
go_mucus_secreting <- enrichGO(gene=mucus_secreting_geneList, OrgDb="org.Hs.eg.db",keyType = "ENSEMBL", ont="BP")
go_mucus_secreting_df <-as.data.frame(go_mucus_secreting)
colnames(go_mucus_secreting_df)
```

## 7. Validation

Compare results of the same cell types from different datasets e.g. cellRef vs. Hlca

Compare results from different cell types. e.g. mucus_secreting cell vs. ionocyte.

Subsampling, Bootstrap Aggregation,Perturbation ( add 10% cells from other cell type)

Compare with the known go term of the cell type.

## 8. Interpretation

## 9. Reporting
Summarize findings in a detailed report or manuscript with figures and statistical analyses.
