---
title: "Data plan for geneExp"
author: "Bingfang"
date: "12/20/2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. Objective

Identify a list of true expressed genes of a specific cell type, which will use to for functional analysis of biological process/pathway of the cell type.

## 2. Set environment


```{r, echo=TRUE,warning=FALSE,message=FALSE}
#install.packages("sads")  # not working on local MAC
library(sads)
#install.packages("BiocParallel")
library(BiocParallel)
library(Seurat)
library(org.Hs.eg.db)
#install.packages("Matrix")
library(Matrix)
# package for goenrichment
library(clusterProfiler)
library(AnnotationDbi)

# Get gene-specific distributions
# Publication:https://academic.oup.com/biostatistics/article/23/4/1150/6622011
# https://github.com/igrabski/scRNAseq-cell-type

load('params_EM_81020.rda') 
pi.all.g <- params[[1]]         # π-null,π-off, probabilty of null and off state.    
g.on.all.g <- params[[2]]       # σ1j⁠
g.off.all.g <- params[[3]]      # σ0j
a.all.g <- params[[4]]          # αj⁠
sigma.all.g <- params[[5]]      # μ0j,μ1j shift from (μ0 , μ1)
mu.g<-c(-12.73153,-8.301349)    # global means for gene rates on a log scale, (μ0 , μ1)

```

## 2. Data collection and preprocess.

### 2.1 Read .rds file

HLCA data from cellXgene

cellRef data from cellXgene

The authors of cellRef match "celltype_level3_fullname" with "putative CL label" and "CL term"

```{r, results='hide'} 

hlca = readRDS("/gpfs/gsfs10/users/xubr/cellXgene/scRNAseq-cell-type/data/b351804c-293e-4aeb-9c4c-043db67f4540_hlca_core.rds")
cellRef = readRDS("/gpfs/gsfs10/users/xubr/cellXgene/scRNAseq-cell-type/data/443f7fb8-2a27-47c3-98f6-6a603c7a294e_cellRef.rds")


summary(hlca@assays$RNA@meta.features$feature_type)
head(hlca@assays$RNA@meta.features$feature_name)
table(hlca@assays$RNA@meta.features$feature_is_filtered)
table(hlca@meta.data$cell_type_ontology_term_id)
table(hlca@meta.data$cell_type)
table(hlca@meta.data$ann_finest_level)

summary(cellRef@assays$RNA@meta.features$feature_type)
head(cellRef@assays$RNA@meta.features$feature_name)  ## lncRNA use Ensemble ID
head(cellRef@assays$RNA@meta.features$origSymbol)    ## lncRNA use symbol
table(cellRef@assays$RNA@meta.features$feature_is_filtered)
table(cellRef@meta.data$cell_type_ontology_term_id)
table(cellRef@meta.data$putative_CL_label)
table(cellRef@meta.data$celltype_level3_fullname)


```

Notes:
cellRef anthors provided 47 cell ontology terms, which match to their 48 celltype_level3 clusters. They also provided putative cell ontology names.

HLCA authors provided 50 cell ontology terms with 50 cell type names, which partially match to their 61 finest-level clusters.

IncRNA ensembl ID may not be stable
e.g. ENSG00000238009 
Ensembl gene ENSG00000238009 is no longer in the database.It has been mapped to 1 current identifier  ENSG00000241860.


### 2.2 Remove low quality cells

```{}
## plot

jpeg(file="HLCA_nFeature.jpeg")
VlnPlot(hlca, features = c("nFeaturess_RNA_detected"))   #,raster=FALSE)
dev.off()

jpeg(file="cellRef_nFeature.jpeg")
VlnPlot(cellRef, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
dev.off()

## filter

## need to convert to gene symbol to filter percent.mt < 5
## pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

hlca_filtered <- subset(hlca, subset = nFeaturess_RNA_detected > 200 & nFeaturess_RNA_detected < 2500)
cellRef_filtered <- subset(cellRef,subset = nFeature_RNA > 200 & nFeature_RNA < 2500)
```

HLCA and cellRef are combined datasets, Likely, they should not be filtered using nFeature.

```{}
## prior and after filter
hlca@assays$RNA@counts@Dim
hlca_filtered@assays$RNA@counts@Dim

cellRef$RNA@counts@Dim
cellRef_filtered@assays$RNA@counts@Dim
```


Notes:

High percentage cells will be filtered if using "nFeature_RNA > 200 & nFeature_RNA < 2500". Likely, the combined dataset should not to performfiltering.


## 3 Functions

### 3.1 calculate rate and three state probabilities

```{r, message=FALSE}

# Train reference data
trainReference <- function(ref,pi.all=pi.all.g,mu=mu.g,a=a.all.g,sigma.all=sigma.all.g,
                           g.on.all=g.on.all.g,g.off.all=g.off.all.g,discrim_only=FALSE) {
  # Subset to genes present in both reference data and knowledge base
  common <- intersect(names(ref),rownames(pi.all))
  if (discrim_only) {
    common <- intersect(common,discrim.g)
  }
  N <- sum(ref)
  ref <- ref[common]
  pi.all2 <- pi.all[common,]
  a.all2 <- a[common]
  sigma.all2 <- sigma.all[common,]
  g.on.all2 <- g.on.all[common]
  g.off.all2 <- g.off.all[common]
  
  # Compute mixing probability for each component
  prob.exp <- pi.all2[,1]*dnbinom(ref,1,1/((N/(a.all2))+1))
  prob.ln1 <- sapply(common,function(j) 
    pi.all2[j,2]*sads::dpoilog(ref[j],mu[1]+g.off.all2[j]+log(N),sigma.all2[j,1]))
  prob.ln2 <- sapply(common,function(j) 
    (1-pi.all2[j,1]-pi.all2[j,2])*sads::dpoilog(ref[j],mu[2]+g.on.all2[j]+log(N),sigma.all2[j,2])) 
  return(data.frame(rate=ref/N,exp=prob.exp/(prob.exp+prob.ln1+prob.ln2),ln1=prob.ln1/(prob.exp+prob.ln1+prob.ln2),ln2=prob.ln2/(prob.exp+prob.ln1+prob.ln2)))
}


# Train all reference
trainAllReference <- function(data,labels,pi.all=pi.all.g,mu=mu.g,a=a.all.g,sigma.all=sigma.all.g,
                              g.on.all=g.on.all.g,g.off.all=g.off.all.g,discrim_only=FALSE) {
  cell.types <- unique(labels)
  dataSums <- lapply(cell.types,function(c) rowSums(data[,labels==c]))

  d.list <- bplapply(dataSums,trainReference,pi.all=pi.all,mu=mu,a=a,sigma.all=sigma.all,
                     g.on.all=g.on.all,g.off.all=g.off.all,discrim_only=discrim_only)
  names(d.list) <- cell.types
  return(d.list)
}

```

### getBarcode

```{r}
getBarcode<-function(d.list) {
  barcode <- sapply(d.list,function(x) 1-x[,2]-x[,3])      # ### Barcode is   ln2=prob.ln2/(prob.exp+prob.ln1+prob.ln2) if greater than 0
  barcode[barcode<0] <- 0
  colnames(barcode) <- names(d.list)
  rownames(barcode) <- rownames(d.list[[1]])
  return(barcode)
}
```

### Run by cell type

```{r}

## for cellRef, meta_feature is "celltype_level3_fullname"
generateCellType_barcode <- function(seurat_dataset, cellType, cut_off){
  # sebset by cell type
  data_subset <- subset(seurat_dataset, subset = celltype_level3_fullname == cellType)
  matrix<- as.matrix(data_subset@assays$RNA@counts)
  dim(matrix)[2]
  labels<-c(rep(cell_type,dim(matrix)[2]))
  length(labels)
  
  # calculte rate and probability
  data.d <- trainAllReference(matrix,labels,discrim_only=FALSE)

  # get barcode and gene list
  barcodes <- getBarcode(data.d)
  geneList <-rownames(barcodes)[barcodes > cut_off]
  length(geneList)
  write.table(geneList, file = paste0(cellType, cut_off, ".tsv"), sep = "\n", row.names = FALSE, col.names = FALSE, quote = FALSE)
  return(barcodes)
  
}

unique(cellRef$celltype_level3_fullname)
uNique(cellRef$putative_CL_label)
barcode_dic={}
seurat_dataset<-cellRef
cellType<-"Alveolar fibroblast 2"
cell_type <-gsub(" ","_",cellType)
cut_off<-0.75
code<-generateCellType_barcode(cellRef, cellType, cut_off)
barcode_dic$cell_type<-as.data.frame(code) 
head(barcode_dic$cell_type)

df_8marker<-paste("id" , "barcode" , "cell_type",sep="\t")
for(id in c("ENSG00000038945","ENSG00000170323", "ENSG00000183019","ENSG00000102575","ENSG00000131471","ENSG00000139329","ENSG00000136546","ENSG00000077943")){
    line<-paste(id, barcode_dic$cell_type[id,],cell_type,sep="\t")
    df_8marker<-rbind(df_8marker,line)
  
}


```


## loop through multiple cell types


```{}

runBatch<-function(seurat_dataset,cell_type_list,cut_off_list){
  dic ={}
  for(cellType in cell_type_list){
   
    for (cut_off in cut_off_list) {
      
      code<-generateCellType_barcode(seurat_dataset, cellType, cut_off)
      print(head(code))
      cell_type <-paste0(gsub(" ","_",cellType), "_",cut_off)
      print(cell_type)
      dic$cell_type<-as.data.frame(code) 
    }
  }
  return(dic)
}

cell_type_list <- c("Alveolar macrophage","Alveolar fibroblast2")
cut_off_list <- c(0.75,0.95)
barcode_dic1<-runBatch(cellRef,cell_type_list,cut_off_list)



```

## 6. Visualization and validation of the gene expression list

1. check probability of NS-forest marker gene expression

                     HLCA_NSForestMarkers	CellRef_NSForestMarkers
Alveolar macrophages ['MSR1', 'FABP4']	['MCEMP1', 'ACP5']
Alveolar Fibroblasts ['AOC3', 'LUM']	['SCN7A', 'ITGA8']

MSR1 ENSG00000038945 
FABP4 ENSG00000170323 
MCEMP1 ENSG00000183019
ACP5 ENSG00000102575

AOC3 ENSG00000131471 
LUM ENSG00000139329
SCN7A ENSG00000136546
ITGA8 ENSG00000077943 


MSR1 ENSG00000038945
hlca_2alveolar.d[[1]]["ENSG00000038945",] # 97% chance that MSR1 is expresed in Alveolar macrophages
hlca_2alveolar.d[[2]]["ENSG00000038945",] # 97% chance that MSR1 is expresed in Alveolar Fibroblasts
 
cellRef_2alveolar.d[[1]]["ENSG00000038945",] # 91% chance that MSR1 is expresed in Alveolar macrophages
cellRef_2alveolar.d[[2]]["ENSG00000038945",] # 22% chance that MSR1 is expresed in Alveolar Fibroblasts



### Probability of 4 Alveolar Fibroblasts markers detected in Alveolar Fibroblasts cluster in hlca and cellRef

```{r}

df_fibroblast_marker<-hlca_2alveolar.d[[1]][1,]
for(id in c("ENSG00000131471","ENSG00000139329","ENSG00000136546","ENSG00000077943")){
    
    df_fibroblast_marker<-rbind(df_fibroblast_marker,hlca_Alveolar_fibroblasts.d$`Alveolar fibroblasts`[id,],cellRef_2alveolar.d$`Alveolar fibroblast2`[id,]) 
}

for(id in c("ENSG00000131471","ENSG00000139329","ENSG00000136546","ENSG00000077943")){
    #df_fibroblast_marker$cluster<-c("Alveolar macrophages")
    df_fibroblast_marker<-rbind(df_fibroblast_marker,hlca_Alveolar_macrophages.d$`Alveolar macrophages`[id,],cellRef_2alveolar.d$`Alveolar macrophage`[id,]) 
}
df_fibroblast_marker <- df_fibroblast_marker[2:17,]
df_fibroblast_marker

```

### Probability of 4 Alveolar macrophages markers detected in Alveolar macrophages cluster and in Alveolar fibroblasts

```{r}


marker_Alveolar_macrophages<-c("ENSG00000038945","ENSG00000170323", "ENSG00000183019","ENSG00000102575")
df_macrophages_marker<-hlca_2alveolar.d[[1]][1,]
for(id in marker_Alveolar_macrophages){
    
    #df_fibroblast_marker$cluster<-c("Alveolar macrophages")
    df_macrophages_marker<-rbind(df_macrophages_marker,hlca_2alveolar.d$`Alveolar macrophages`[id,],cellRef_2alveolar.d$`Alveolar macrophage`[id,]) 
}

for(id in marker_Alveolar_macrophages){
    df_macrophages_marker<-rbind(df_macrophages_marker,hlca_2alveolar.d$`Alveolar fibroblasts`[id,],cellRef_2alveolar.d$`Alveolar fibroblast2`[id,]) 
}


df_macrophages_marker <- df_macrophages_marker[2:17,]
df_macrophages_marker


```

df_8marker_in_macrophages<-hlca_Alveolar_macrophages.d[[1]][1,]
for(id in c("ENSG00000038945","ENSG00000170323", "ENSG00000183019","ENSG00000102575","ENSG00000131471","ENSG00000139329","ENSG00000136546","ENSG00000077943")){
    
    df_8marker_in_macrophages<-rbind(df_8marker_in_macrophages,hlca_Alveolar_macrophages.d$`Alveolar macrophages`[id,])
  
}
print(df_8marker_in_macrophages[2:9,])


df_8marker_in_fibroblast<-hlca_Alveolar_macrophages.d[[1]][1,]
for(id in c("ENSG00000038945","ENSG00000170323", "ENSG00000183019","ENSG00000102575","ENSG00000131471","ENSG00000139329","ENSG00000136546","ENSG00000077943")){
    
    df_8marker_in_fibroblast<-rbind(df_8marker_in_fibroblast,hlca_Alveolar_fibroblasts.d$`Alveolar fibroblasts`[id,])
  
}
print(df_8marker_in_fibroblast[2:9,])

```{r}
library(clusterProfiler)
mucus_secreting_geneList
go_mucus_secreting <- enrichGO(gene=mucus_secreting_geneList, OrgDb="org.Hs.eg.db",keyType = "ENSEMBL", ont="BP")
go_mucus_secreting_df <-as.data.frame(go_mucus_secreting)
colnames(go_mucus_secreting_df)
```

## 7. Validation

Compare results of the same cell types from different datasets e.g. cellRef vs. Hlca

Compare results from different cell types. e.g. mucus_secreting cell vs. ionocyte.

Subsampling, Bootstrap Aggregation,Perturbation ( add 10% cells from other cell type)

Compare with the known go term of the cell type.

## 8. Interpretation

## 9. Reporting
Summarize findings in a detailed report or manuscript with figures and statistical analyses.
